### ISSUE TITLE
ImFontAtlas::AddFontFromFileTTF now requires context

### ISSUE THREAD

**Issue by tksuoran** [ocornut/imgui#7314]
### Version/Branch of Dear ImGui:

Version 1.90.2, Branch: docking

### Back-ends:

custom (erhe)

### Compiler, OS:

WIndows 11 + MSVC 2022

### Full config/build information:

```
Dear ImGui 1.90.2 (19020)
--------------------------------
sizeof(size_t): 8, sizeof(ImDrawIdx): 2, sizeof(ImDrawVert): 20
define: __cplusplus=202002
define: IMGUI_DISABLE_OBSOLETE_FUNCTIONS
define: IMGUI_DISABLE_OBSOLETE_KEYIO
define: _WIN32
define: _WIN64
define: _MSC_VER=1938
define: _MSVC_LANG=202002
define: IMGUI_HAS_VIEWPORT
define: IMGUI_HAS_DOCK
--------------------------------
io.BackendPlatformName: imgui_impl_erhe
io.BackendRendererName: erhe
io.ConfigFlags: 0x00000049
 NavEnableKeyboard
 NavNoCaptureKeyboard
 DockingEnable
io.ConfigViewportsNoDecoration
io.ConfigInputTextCursorBlink
io.ConfigWindowsResizeFromEdges
io.ConfigWindowsMoveFromTitleBarOnly
io.ConfigMemoryCompactTimer = 60.0
io.BackendFlags: 0x0000000A
 HasMouseCursors
 RendererHasVtxOffset
--------------------------------
io.Fonts: 4 fonts, Flags: 0x00000000, TexSize: 512,256
io.DisplaySize: 1920.00,1080.00
io.DisplayFramebufferScale: 1.00,1.00
--------------------------------
style.WindowPadding: 2.00,2.00
style.WindowBorderSize: 0.00
style.FramePadding: 2.00,2.00
style.FrameRounding: 3.00
style.FrameBorderSize: 0.00
style.ItemSpacing: 2.00,2.00
style.ItemInnerSpacing: 2.00,2.00
```

### Details:

**My Issue/Question:**

I am upgrading erhe from v1.89.6 to v1.90.2.

I share fonts with multiple contexts. In 1.89.6, I was able to create font atlas before creating context. This no longer seems to be possible in v1.90.2 because:

- `ImFontAtlas::AddFontFromFileTTF()` calls `ImFileLoadToMemory()`
- `ImFileLoadToMemory()` calls `ImFileOpen()`
- `ImFileOpen()` assumes there is current `ImGuiContext`, crashing if this is not the case

Is it reasonable expectation that I should be able to create font atlases before contexts, so that I can share that atlas to all contexts that I then later create?

**Comment by tksuoran** [comment]
Disabling the Windows specific code path in `ImFileOpen()` works around the issue.

**Comment by PathogenDavid** [comment]
For reference, this was caused by https://github.com/ocornut/imgui/commit/daf49e9d829bde0bd3d5124139e9191a54521ae6

> Is it reasonable expectation that I should be able to create font atlases before contexts, so that I can share that atlas to all contexts that I then later create?

According to the documentation on `CreateContext` this should be supported, so yes this is a reasonable expectation:

https://github.com/ocornut/imgui/blob/f966da1f8fb569b99dd4220162f2d9bde6f3b153/imgui.h#L295C22-L295C93

I created a PR to fall back on a local buffer when the context isn't available yet: https://github.com/ocornut/imgui/pull/7315

**Comment by ocornut** [comment]
> Is it reasonable expectation that I should be able to create font atlases before contexts, so that I can share that atlas to all contexts that I then later create?

Thanks for reporting!
This wasn't really designed per-se, but I think it makes sense that we make this possible. I see that David already made a PR for it. 
I need to log off now but I'll think about this a little more tomorrow.

(I've been meaning to tag another release as one meaningful popup bug was fixed too, so perhaps there's fix a release with fix tomorrow.)

**Comment by ocornut** [comment]
> According to the documentation on CreateContext this should be supported, so yes this is a reasonable expectation:

Thanks for digging it.

> I created a PR to fall back on a local buffer when the context isn't available yet: https://github.com/ocornut/imgui/pull/7315

I had to rework this: another sort-of-unsaid expectation is that the loose ImXXX functions shouldn't depend on the current dear imgui context. While your PR had a check for the existence of a context, this would potentially break if we start to consider use in e.g. multi-threaded scenario. I reworked it as 9dfa239, using a backup stack buffer (essentially a small vector pattern, which I might introduce properly in imgui_internal.h later as this is a common need).

Thanks David and Timo!
